### ISSUE TITLE
Download of multiple documents

### ISSUE THREAD

**Issue by najiba-f** [k-box/k-box#423]
### Short Description of the issue
Currently, it is only possible to download one document

### Expected behavior
It should be possible to select several documents and download it as a zip file.

**Comment by avvertix** [comment]
Yes, that would be a good option to add. The only point to reflect is if the zip file should be generated on the flight and downloaded immediately when the user clicks the button, or if the user should be notified once the zip file is ready.

**Comment by shenriod** [comment]
Just a thought: this is not only about downloading. It would mostly be great to allow sharing of several documents with one single link. Will that be part of this PR?

**Comment by avvertix** [comment]
@shenriod probably not, this will focus on the download only

**Comment by najiba-f** [comment]
Due to the potential issues may caused by a user requesting an excessively large download file,  I suggest that we either:
- Put a limit on file size (zip). For example, 200 mb per download  
- Or put the task on background to first generate the zip file then send it out to the user's email
Are there any other suggestion on how the whole bulk download process should go?

**Comment by avvertix** [comment]
> Put a limit on file size (zip). For example, 200 mb per download

I'm not sure this is suitable as if a video file is selected it will become more difficult to split into multiple zip files or to block the option

> Or put the task on background to first generate the zip file then send it out to the user's email

This could be a better option in case lot of files are selected or the selected ones are large. Of course the user will be notified with a link to zip file ready for download.

From the perspective of user experience I guess this "notify when ready" option is better than eventually getting no zip file at all or errors due to the timeout. Unless you have insights from your user I would go into the asynchronous creation of the zip file and then notify the user when ready for download.

**Comment by shenriod** [comment]
Just to add my perspective as a user (with little understanding of the technical complexity of implementing it):

The option with sending a DL link per email is fully ok, if I try to download a big packet of files.

But if I just want to quickly download 5 pdfs from a collection, this will be a little bit frustrating.

Would it be possible to implement both options? If I want to download less than X files (or less than Y Mb), then the zip is directly generated and downloaded. If I want more than X files (or more than Y Mb), then I get a message that this is too big and that I will receive an email with a DL link as soon as the zip file is ready

**Comment by najiba-f** [comment]
@shenriod  I think that could be possible.
@avvertix  should we put limitation on the size, or the user can download any size? For example, 1TB or 2TB.

**Comment by avvertix** [comment]
That's a good question @najiba-f. From the technical perspective a zip file has no upper limit, but file systems do have it. For example the old FAT32 format has 4 GB limit. I saw it still on some usb keys or external disks. The second limitation/problem might be the time required to create the zip archive, which is difficult to estimate and this might cause the background job to be stopped just moments before having a complete zip file.

So my perspective would be to limit to 4GB the maximum size of the zip file and it cannot contain more than 65000 documents. This limit should be configurable or easy to change so we can extend in the future after observing how the feature was used and how the system behaved.

**Comment by shenriod** [comment]
But what happens if the limit is surpassed? Should the user get an error message (asking him to choose less files) or should his files be split into several zip archives?

**Comment by najiba-f** [comment]
@avvertix  got it. 
@shenriod I think the user should be notified or it would be better if we summarize the data size before compressing it. So, we won't let user to select more files. what do you think @avvertix @shenriod?

**Comment by avvertix** [comment]
I would indeed check the size of the documents to be included in the zip in the first place and then tell the user what to expect. Considering the way file transfer works we need to present eventual messages to the user in case of errors. For example if the request for the download exceed the timeout of if the user change the page the browser might abort the request.

@najiba-f considering the messages in the chat we had, the library https://github.com/maennchen/ZipStream-PHP and the [streamed response](https://laravel.com/docs/7.x/responses#file-downloads) (which if I got correctly is baked by Symfony's [StreamedResponse](https://symfony.com/doc/current/components/http_foundation.html#streaming-a-response)) could be a viable option for the first implementation to observe the behavior and eventual connection/timeout problems.

**Comment by shenriod** [comment]
> I would indeed check the size of the documents to be included in the zip in the first place and then tell the user what to expect. 

I will maybe say something stupid, but how can we know it before actually compressing the files? Depending on the type of files, the compression rate will be very different, so we can probably not know _a priori_ whether the zip file will be smaller of bigger than 4 Gb. So, in other words, we need to actually compress the files before we know what the output will look like. Or am I fully wrong here?

**Comment by najiba-f** [comment]
@avvertix  Okay, on it.
@shenriod As I understood it will be the total size of files to be compressed.

**Comment by avvertix** [comment]
@shenriod @najiba-f yes, we will look at the total size of the files to be compressed, as this is equivalent to the worst case scenario were the compression factor is zero.
